fuzzy/transport.go

```diff
@@ -0,0 +1,322 @@
+	w := bufio.NewWriter(fw)
+	for i := range t.ae {
+		e := &t.ae[i]
+		fmt.Fprintf(w, "%v -> %v\t%8d - %8d : %8d\n", e.source, e.target, e.firstIndex, e.lastIndex, e.commitIndex)
+	}
+	w.Flush()
+	fw.Close()
+}
+
+func firstIndex(a *raft.AppendEntriesRequest) uint64 {
+	if len(a.Entries) == 0 {
+		return 0
+	}
+	return a.Entries[0].Index
+}
+
+func lastIndex(a *raft.AppendEntriesRequest) uint64 {
+	if len(a.Entries) == 0 {
+		return 0
+	}
+	return a.Entries[len(a.Entries)-1].Index
+}
+
+// RequestVote sends the appropriate RPC to the target node.
+func (t *transport) RequestVote(id raft.ServerID, target raft.ServerAddress, args *raft.RequestVoteRequest, resp *raft.RequestVoteResponse) error {
+	return t.sendRPC(string(target), args, resp)
+}
+
+// InstallSnapshot is used to push a snapshot down to a follower. The data is read from
+// the ReadCloser and streamed to the client.
+func (t *transport) InstallSnapshot(id raft.ServerID, target raft.ServerAddress, args *raft.InstallSnapshotRequest, resp *raft.InstallSnapshotResponse, data io.Reader) error {
+	t.log.Printf("INSTALL SNAPSHOT *************************************")
+	return errors.New("huh")
+}
+
+// EncodePeer is used to serialize a peer name.
+func (t *transport) EncodePeer(id raft.ServerID, p raft.ServerAddress) []byte {
+	return []byte(p)
+}
+
+// DecodePeer is used to deserialize a peer name.
+func (t *transport) DecodePeer(p []byte) raft.ServerAddress {
+	return raft.ServerAddress(p)
+}
+
+// SetHeartbeatHandler is used to setup a heartbeat handler
+// as a fast-pass. This is to avoid head-of-line blocking from
+// disk IO. If a Transport does not support this, it can simply
+// ignore the call, and push the heartbeat onto the Consumer channel.
+func (t *transport) SetHeartbeatHandler(cb func(rpc raft.RPC)) {
+}
+
+// AppendEntriesPipeline returns an interface that can be used to pipeline
+// AppendEntries requests.
+func (t *transport) AppendEntriesPipeline(id raft.ServerID, target raft.ServerAddress) (raft.AppendPipeline, error) {
+	p := &pipeline{
+		t:        t,
+		id:       id,
+		target:   target,
+		work:     make(chan *appendEntry, 100),
+		consumer: make(chan raft.AppendFuture, 100),
+	}
+	go p.run()
+	return p, nil
+}
+
+type appendEntry struct {
+	req      *raft.AppendEntriesRequest
+	res      *raft.AppendEntriesResponse
+	start    time.Time
+	err      error
+	ready    chan error
+	consumer chan raft.AppendFuture
+}
+
+func (e *appendEntry) Request() *raft.AppendEntriesRequest {
+	return e.req
+}
+
+func (e *appendEntry) Response() *raft.AppendEntriesResponse {
+	<-e.ready
+	return e.res
+}
+
+func (e *appendEntry) Start() time.Time {
+	return e.start
+}
+
+func (e *appendEntry) Error() error {
+	<-e.ready
+	return e.err
+}
+
+func (e *appendEntry) Respond(err error) {
+	e.err = err
+	close(e.ready)
+	e.consumer <- e
+}
+
+type pipeline struct {
+	t        *transport
+	target   raft.ServerAddress
+	id       raft.ServerID
```

Comment:
Added this id here and modified all interface methods in the fuzzy test to send the id along